File: src/data/items.ts
/**
 * The typed client for the item commands.
 *
 * The host speaks snake_case, because that is what the schema speaks. The
 * interface speaks camelCase. The translation happens exactly once, here,
 * rather than leaking a database naming convention into every component.
 */

import { invoke } from '@tauri-apps/api/core';

import type { Collection, Item } from '@/domain/item';

interface RawItem {
  id: string;
  collection_id: string;
  parent_item_id: string | null;
  title: string;
  position: string;
  completed_at: string | null;
  created_at: string;
  updated_at: string;
}

interface RawCollection {
  id: string;
  name: string;
  icon: string | null;
  color: string | null;
  position: string;
}

function toItem(raw: RawItem): Item {
  return {
    id: raw.id,
    collectionId: raw.collection_id,
    parentItemId: raw.parent_item_id,
    title: raw.title,
    position: raw.position,
    completedAt: raw.completed_at,
    createdAt: raw.created_at,
    updatedAt: raw.updated_at,
  };
}

function toCollection(raw: RawCollection): Collection {
  return {
    id: raw.id,
    name: raw.name,
    icon: raw.icon,
    color: raw.color,
    position: raw.position,
  };
}

export async function listCollections(): Promise<Collection[]> {
  const raw = await invoke<RawCollection[]>('collections_list');
  return raw.map(toCollection);
}

export async function listItems(
  collectionId: string | null,
  includeCompleted: boolean,
): Promise<Item[]> {
  const raw = await invoke<RawItem[]>('items_list', { collectionId, includeCompleted });
  return raw.map(toItem);
}

/**
 * Create an item.
 *
 * `position` is computed by the domain layer and passed in: the ordering
 * algorithm lives in one tested place and the host stores what it is given
 * (ADR-006).
 */
export async function createItem(
  collectionId: string,
  title: string,
  position: string,
): Promise<Item> {
  const raw = await invoke<RawItem>('item_create', {
    item: { collection_id: collectionId, title, position },
  });
  return toItem(raw);
}

export async function setItemCompleted(id: string, completed: boolean): Promise<Item> {
  return toItem(await invoke<RawItem>('item_set_completed', { id, completed }));
}

export async function renameItem(id: string, title: string): Promise<Item> {
  return toItem(await invoke<RawItem>('item_rename', { id, title }));
}

export async function moveItem(
  id: string,
  position: string,
  collectionId: string | null = null,
): Promise<Item> {
  return toItem(await invoke<RawItem>('item_move', { id, position, collectionId }));
}

export async function deleteItem(id: string): Promise<void> {
  await invoke<void>('item_delete', { id });
}

File: src-tauri/src/commands/items.rs
//! Item commands: the typed boundary for the first vertical slice.
//!
//! Each one validates nothing beyond what the repository already checks,
//! delegates, and returns. Ordering keys arrive from the domain layer rather
//! than being computed here, so the algorithm stays in one tested place.

use tauri::State;

use crate::db::models::{CaptureRequest, Collection, Item, ItemPlan, ItemSchedule, NewItem};
use crate::db::{items, Db};
use crate::error::Result;

/// Emitted to every window after a write made outside the main window — the
/// quick-capture window has its own page and its own cache, and the workspace
/// behind it must not look stale to the person who then opens the main one.
pub const WORKSPACE_CHANGED: &str = "workspace:changed";

/// Quick capture: one line, already parsed by the domain layer, written whole.
#[tauri::command]
pub fn item_capture(
    app: tauri::AppHandle,
    db: State<'_, Db>,
    request: CaptureRequest,
) -> Result<Item> {
    let item = {
        let mut conn = db.0.lock().expect("the database lock was poisoned");
        items::capture_item(&mut conn, request)?
    };
    let _ = tauri::Emitter::emit(&app, WORKSPACE_CHANGED, ());
    Ok(item)
}

#[tauri::command]
pub fn collections_list(db: State<'_, Db>) -> Result<Vec<Collection>> {
    let conn = db.0.lock().expect("the database lock was poisoned");
    items::list_collections(&conn)
}

#[tauri::command]
pub fn items_list(
    db: State<'_, Db>,
    collection_id: Option<String>,
    include_completed: bool,
) -> Result<Vec<Item>> {
    let conn = db.0.lock().expect("the database lock was poisoned");
    items::list_items(&conn, collection_id.as_deref(), include_completed)
}

#[tauri::command]
pub fn item_create(db: State<'_, Db>, item: NewItem) -> Result<Item> {
    let mut conn = db.0.lock().expect("the database lock was poisoned");
    items::create_item(&mut conn, item)
}

#[tauri::command]
pub fn item_set_completed(db: State<'_, Db>, id: String, completed: bool) -> Result<Item> {
    let conn = db.0.lock().expect("the database lock was poisoned");
    items::set_completed(&conn, &id, completed)
}

#[tauri::command]
pub fn item_rename(db: State<'_, Db>, id: String, title: String) -> Result<Item> {
    let mut conn = db.0.lock().expect("the database lock was poisoned");
    items::rename_item(&mut conn, &id, &title)
}

#[tauri::command]
pub fn item_move(
    db: State<'_, Db>,
    id: String,
    position: String,
    collection_id: Option<String>,
) -> Result<Item> {
    let conn = db.0.lock().expect("the database lock was poisoned");
    items::move_item(&conn, &id, &position, collection_id.as_deref())
}

#[tauri::command]
pub fn item_delete(db: State<'_, Db>, id: String) -> Result<()> {
    let mut conn = db.0.lock().expect("the database lock was poisoned");
    items::delete_item(&mut conn, &id)
}

/// Move a card on a board: position and grouping field, in one transaction.
#[tauri::command]
pub fn item_move_on_board(
    db: State<'_, Db>,
    id: String,
    position: String,
    property_id: Option<String>,
    value: serde_json::Value,
) -> Result<Item> {
    let mut conn = db.0.lock().expect("the database lock was poisoned");
    items::move_on_board(&mut conn, &id, &position, property_id.as_deref(), &value)
}

/// Set an item's dates and repetition.
#[tauri::command]
pub fn item_set_schedule(db: State<'_, Db>, id: String, schedule: ItemSchedule) -> Result<Item> {
    let conn = db.0.lock().expect("the database lock was poisoned");
    items::set_schedule(&conn, &id, schedule)
}

/// Complete one occurrence of a repeating item, moving it to its next date.
#[tauri::command]
pub fn item_complete_occurrence(
    db: State<'_, Db>,
    id: String,
    next_due_at: Option<String>,
) -> Result<Item> {
    let mut conn = db.0.lock().expect("the database lock was poisoned");
    items::complete_occurrence(&mut conn, &id, next_due_at.as_deref())
}

/// How long this task takes, and whether it takes any time at all.
#[tauri::command]
pub fn item_set_plan(db: State<'_, Db>, id: String, plan: ItemPlan) -> Result<Item> {
    let conn = db.0.lock().expect("the database lock was poisoned");
    items::set_plan(&conn, &id, plan)
}
